In [119]:
#PANDAS
#PANDAS
#PANDAS

PANDAS
PANDAS
PANDAS

PANDAS SERİES

Pandas serileri, tek boyutlu ve index bilgisi barındıran bir veri tipidir

Pandas data frame'i ise, çok boyutlu ve index bilgisi barındıran bir diğer veri tipidir

pd.Series() #Liste gibi bir veri ver ve ben bunu PANDAS serisi olarak tanımlayayım... DER...

In [120]:
import pandas as pd

pd.Series([10,77,12,4,5]) #Liste gibi bir veri ver ve ben bunu PANDAS serisi olarak tanımlayayım... DER...

0    10
1    77
2    12
3     4
4     5
dtype: int64

In [121]:
import pandas as pd

s = pd.Series([10,77,12,4,5])
type(s)

pandas.core.series.Series

In [122]:
s.index #indexlerini verir

RangeIndex(start=0, stop=5, step=1)

In [123]:
s.dtype #veri tipini verir

dtype('int64')

In [124]:
s.size #kaç elemanlı olduğunu verir

5

In [125]:
s.ndim #boyutunu verir

1

In [126]:
s.values #değerlerini verir (numpy arrayi) 

array([10, 77, 12,  4,  5], dtype=int64)

In [127]:
s.head()  #ilk 5 elemanı verir  (varsayılan olarak 5 elemanı verir)  s.head(3) 3 elemanı verir

0    10
1    77
2    12
3     4
4     5
dtype: int64

In [128]:
s.head(3) # ilk 3 elemanı verir

0    10
1    77
2    12
dtype: int64

In [129]:
s.tail() #son 5 elemanı verir

0    10
1    77
2    12
3     4
4     5
dtype: int64

In [130]:
s.tail(2) #son 3 elemanı verir

3    4
4    5
dtype: int64

# VERİ OKUMA

In [131]:
import pandas as pd
df = pd.read_csv("advertising.csv") #csv dosyasını oku ve df değişkenine ata

df.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9


AŞAĞIDA, .read_csv ' nin nereden geldiğinin bilgisini paylaştım

In [132]:
from pandas.io.api import (
    # excel
    ExcelFile,
    ExcelWriter,
    read_excel,
    # parsers
    read_csv,
    read_fwf,
    read_table,
    # pickle
    read_pickle,
    to_pickle,
    # pytables
    HDFStore,
    read_hdf,
    # sql
    read_sql,
    read_sql_query,
    read_sql_table,
    # misc
    read_clipboard,
    read_parquet,
    read_orc,
    read_feather,
    read_gbq,
    read_html,
    read_xml,
    read_json,
    read_stata,
    read_sas,
    read_spss,
)

In [133]:
# VERİYE HIZLI BAKIŞ

# Veriye hızlı bakış

In [134]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic") #seaborn kütüphanesinden veri seti yükle (varsayılan olarak iris veri seti yüklenir) 
df.head() #ilk 5 satırı göster

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [135]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.tail() #son 5 satırı göster

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

In [136]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.shape #satır ve sütun sayısını verir

(891, 15)

In [137]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.info() #veri seti hakkında bilgi verir

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [138]:
# 891 tane tam gözlem varmış
# kaç tane boş olmayan değer varmış (non-null) 
# veri setindeki değişkenlerin veri tipleri nelermiş
# bellekte ne kadar yer kaplıyormuş

# 714 tane tam gözlem varmış
# object ve category
# int64 ve float64
# 64 bitlik yer kaplıyorlar

# category ve object arasındaki fark:
# category değişkenlerin içindeki kategorik değerlerin sayısı az ise category olarak tanımlanır
# object değişkenlerin içindeki kategorik değerlerin sayısı fazla ise object olarak tanımlanır
# yani özel olarak:
# object değişkenlerin içindeki kategorik değerlerin sayısı az ise category olarak tanımlanır

# bizim için objectler de kategorik değişkenlerdir



In [139]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.columns #değişken isimlerini verir

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [140]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index #indexleri verir

RangeIndex(start=0, stop=891, step=1)

In [141]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.describe() #sayısal değişkenlerin istatistiksel özetini verir

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [142]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.describe().T #transpozunu alır 
# Transpoz = satırlar sütunlar olur, sütunlar satırlar olur

count       mean        std   min      25%      50%   75%       max
survived  891.0   0.383838   0.486592  0.00   0.0000   0.0000   1.0    1.0000
pclass    891.0   2.308642   0.836071  1.00   2.0000   3.0000   3.0    3.0000
age       714.0  29.699118  14.526497  0.42  20.1250  28.0000  38.0   80.0000
sibsp     891.0   0.523008   1.102743  0.00   0.0000   0.0000   1.0    8.0000
parch     891.0   0.381594   0.806057  0.00   0.0000   0.0000   0.0    6.0000
fare      891.0  32.204208  49.693429  0.00   7.9104  14.4542  31.0  512.3292

In [143]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.isnull().values.any() #veri setinde boş değer var mı yok mu? True False döner
# True dönerse boş değer var demektir
# False dönerse boş değer yok demektir

True

In [144]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.isnull() #veri setindeki boş değerleri gösterir (True False döner) 
# True dönerse boş değer var demektir
# False dönerse boş değer yok demektir

# True olanlar boş değerdir

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

In [145]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.isnull().values #veri setindeki boş değerleri gösterir (True False döner)
# True dönerse boş değer var demektir
# False dönerse boş değer yok demektir

# True olanlar boş değerdir

array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])

In [146]:
#df.isnull() ile df.isnull.values arasındaki şudur:
#df.isnull() ile True False döner (boş değer var mı yok mu?)
#df.isnull.values ile True False döner (boş değer var mı yok mu?)

#isnull() 'un değeri DataFrame'dir, values'ın değeri numpy arrayidir


In [147]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.isnull().values.any() #veri setinde boş değer var mı yok mu? True False döner
# True dönerse boş değer var demektir

True

In [148]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.isnull().sum() #boş değerlerin toplam sayısını verir (her değişken için) 
# ne yaptı ? 
# her değişken için boş değer var mı yok mu? True False döner
# True dönerse boş değer var demektir
# False dönerse boş değer yok demektir
# True olanları toplar ve toplam sayıyı verir (her değişken için) 

# HANGİ DEĞİŞKENDEN KAÇ TANE BOŞ DEĞER VAR?


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

###### survived         0
###### pclass           0
###### sex              0
###### age            177
###### sibsp            0
###### parch            0
###### fare             0
###### embarked         2
###### class            0
###### who              0
###### adult_male       0
###### deck           688
###### embark_town      2
###### alive            0
###### alone            0

###### buradaki kategorik değişkenlerden birisinin mesela "sex" 
###### bunun içerisinde kaç tane sınıf olduğu ve 
###### ve bu sınıfların kaçar tane olduğu bilgisine erişmek istiyoruz
###### yanii
###### kadından kaç tane var, erkekten kaç tane var...


In [149]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df["sex"] #sex değişkenini verir (seri olarak) şunu yaptık :
#  

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: object

In [150]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df["sex"].head() #ilk 5 satırını verir (seri olarak)

0      male
1    female
2    female
3    female
4      male
Name: sex, dtype: object

In [151]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df["sex"].value_counts() #sex değişkenindeki kategorik değerlerin sayısını verir


male      577
female    314
Name: sex, dtype: int64

# PANDAS SEÇİM İŞLEMLERİ

In [152]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic") #veri setini yükle ve df değişkenine ata 
df.head() #ilk 5 satırı verir (dataframe olarak)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [153]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index #indexleri verir (satır numaralarını verir) 

RangeIndex(start=0, stop=891, step=1)

In [154]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df[0:13] #0. satırdan 13. satıra kadar olan satırları verir (13 dahil değil)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0          0       3    male  22.0      1      0   7.2500        S   Third   
1          1       1  female  38.0      1      0  71.2833        C   First   
2          1       3  female  26.0      0      0   7.9250        S   Third   
3          1       1  female  35.0      1      0  53.1000        S   First   
4          0       3    male  35.0      0      0   8.0500        S   Third   
5          0       3    male   NaN      0      0   8.4583        Q   Third   
6          0       1    male  54.0      0      0  51.8625        S   First   
7          0       3    male   2.0      3      1  21.0750        S   Third   
8          1       3  female  27.0      0      2  11.1333        S   Third   
9          1       2  female  14.0      1      0  30.0708        C  Second   
10         1       3  female   4.0      1      1  16.7000        S   Third   
11         1       1  female  58.0      0      0  26.5500        S   First   
12         0       3    male  20.0      0      0   8.0500        S   Third   

      who  adult_male deck  embark_town alive  alone  
0     man        True  NaN  Southampton    no  False  
1   woman       False    C    Cherbourg   yes  False  
2   woman       False  NaN  Southampton   yes   True  
3   woman       False    C  Southampton   yes  False  
4     man        True  NaN  Southampton    no   True  
5     man        True  NaN   Queenstown    no   True  
6     man        True    E  Southampton    no   True  
7   child       False  NaN  Southampton    no  False  
8   woman       False  NaN  Southampton   yes  False  
9   child       False  NaN    Cherbourg   yes  False  
10  child       False    G  Southampton   yes  False  
11  woman       False    C  Southampton   yes   True  
12    man        True  NaN  Southampton    no   True

In [155]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.drop #satır veya sütun silmek için kullanılır



<bound method DataFrame.drop of      survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_m

In [156]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.drop()

ValueError: Need to specify at least one of 'labels', 'index' or 'columns'

df.drop Pandas kütüphanesinde DataFrame'lerde kullanılan bir işlemdir. DataFrame, Pandas kütüphanesinin temel veri yapısıdır ve verileri düzenlemek, filtrelemek ve dönüştürmek için kullanılır. df.drop metodu, DataFrame'den belirtilen satırları veya sütunları çıkarmak için kullanılır.

Genel sözdizimi şu şekildedir:
```python
df.drop(labels, axis=0, inplace=False)
```

Burada:

##### labels: Silinmesi istenen satır veya sütunların etiketleri (isimleri) veya etiketlerin listesi olabilir.
##### axis: Eksiltmenin satırları (axis=0) veya sütunları (axis=1) etkileyeceğini belirtmek için kullanılır.
##### inplace: Varsayılan olarak False değeri alır. Eğer True olarak ayarlanırsa, işlem DataFrame'i değiştirir ve DataFrame kendini değiştirir. False olarak bırakırsanız, işlem sonucunda yeni bir DataFrame döndürür ve orijinal DataFrame'i değiştirmez.
Örnek kullanım:

```python
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40],
    'City': ['New York', 'Paris', 'London', 'Tokyo']
}

df = pd.DataFrame(data)

# DataFrame'den bir sütunu çıkaralım
df_dropped = df.drop('City', axis=1)

# DataFrame'den bir satırı çıkaralım
df_dropped = df.drop(2, axis=0)
```

In [ ]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40],
    'City': ['New York', 'Paris', 'London', 'Tokyo']
}

df = pd.DataFrame(data)
df

Name  Age      City
0    Alice   25  New York
1      Bob   30     Paris
2  Charlie   35    London
3    David   40     Tokyo

In [ ]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40],
    'City': ['New York', 'Paris', 'London', 'Tokyo']
}

df = pd.DataFrame(data)
df_dropped = df.drop('City', axis=1) #City değişkenini sil (axis=1 sütun demektir)
df_dropped

Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35
3    David   40

In [ ]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40],
    'City': ['New York', 'Paris', 'London', 'Tokyo']
}

df = pd.DataFrame(data)
df_dropped = df.drop(2, axis=0) #2. satırı sil (axis=0 satır demektir)
df_dropped

Name  Age      City
0  Alice   25  New York
1    Bob   30     Paris
3  David   40     Tokyo

# Birden fazla satır indisini silmek için

In [ ]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40],
    'City': ['New York', 'Paris', 'London', 'Tokyo']
}

df = pd.DataFrame(data)
delete_indexes = [0, 1, 2]

df_dropped = df.drop(delete_indexes, axis=0) #0, 1, 2. satırları sil (axis=0 satır demektir)
df_dropped

Name  Age   City
3  David   40  Tokyo

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df[0:13]


survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0          0       3    male  22.0      1      0   7.2500        S   Third   
1          1       1  female  38.0      1      0  71.2833        C   First   
2          1       3  female  26.0      0      0   7.9250        S   Third   
3          1       1  female  35.0      1      0  53.1000        S   First   
4          0       3    male  35.0      0      0   8.0500        S   Third   
5          0       3    male   NaN      0      0   8.4583        Q   Third   
6          0       1    male  54.0      0      0  51.8625        S   First   
7          0       3    male   2.0      3      1  21.0750        S   Third   
8          1       3  female  27.0      0      2  11.1333        S   Third   
9          1       2  female  14.0      1      0  30.0708        C  Second   
10         1       3  female   4.0      1      1  16.7000        S   Third   
11         1       1  female  58.0      0      0  26.5500        S   First   
12         0       3    male  20.0      0      0   8.0500        S   Third   

      who  adult_male deck  embark_town alive  alone  
0     man        True  NaN  Southampton    no  False  
1   woman       False    C    Cherbourg   yes  False  
2   woman       False  NaN  Southampton   yes   True  
3   woman       False    C  Southampton   yes  False  
4     man        True  NaN  Southampton    no   True  
5     man        True  NaN   Queenstown    no   True  
6     man        True    E  Southampton    no   True  
7   child       False  NaN  Southampton    no  False  
8   woman       False  NaN  Southampton   yes  False  
9   child       False  NaN    Cherbourg   yes  False  
10  child       False    G  Southampton   yes  False  
11  woman       False    C  Southampton   yes   True  
12    man        True  NaN  Southampton    no   True

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
delete_indexes = [1,3,5,7]
df.drop(delete_indexes, axis=0).head(10) #1,3,5,7. satırları sil (axis=0 satır demektir)
# head(10) ilk 10 satırı verir


survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0          0       3    male  22.0      1      0   7.2500        S   Third   
2          1       3  female  26.0      0      0   7.9250        S   Third   
4          0       3    male  35.0      0      0   8.0500        S   Third   
6          0       1    male  54.0      0      0  51.8625        S   First   
8          1       3  female  27.0      0      2  11.1333        S   Third   
9          1       2  female  14.0      1      0  30.0708        C  Second   
10         1       3  female   4.0      1      1  16.7000        S   Third   
11         1       1  female  58.0      0      0  26.5500        S   First   
12         0       3    male  20.0      0      0   8.0500        S   Third   
13         0       3    male  39.0      1      5  31.2750        S   Third   

      who  adult_male deck  embark_town alive  alone  
0     man        True  NaN  Southampton    no  False  
2   woman       False  NaN  Southampton   yes   True  
4     man        True  NaN  Southampton    no   True  
6     man        True    E  Southampton    no   True  
8   woman       False  NaN  Southampton   yes  False  
9   child       False  NaN    Cherbourg   yes  False  
10  child       False    G  Southampton   yes  False  
11  woman       False    C  Southampton   yes   True  
12    man        True  NaN  Southampton    no   True  
13    man        True  NaN  Southampton    no  False

## df.drop(0,axis=0 , inplace=True) 
# inplace True dersek df değişkeni üzerinde değişiklik yapar
### yani değişiklikleri bir yere aktarmadan kaydeder
### inplace False dersek df değişkeni üzerinde değişiklik yapmaz

# değişkeni index' çevirmek
# değişkeni index' çevirmek
# değişkeni index' çevirmek
# değişkeni index' çevirmek

# Elimizdeki değişkeni index'e çevireceğiz
# Yaş değişkenini index'e
# index deki değeri de , bir değişken olarak, değişkenlerin arasına, almak istediğimizi düşünelim

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df["age"].head() #age değişkenini verir (seri olarak)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.age.head() #age değişkenini verir (seri olarak)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index

RangeIndex(start=0, stop=891, step=1)

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"] #indexleri age değişkenine eşitle 
# (age değişkeni index olur) 
df.index

Float64Index([22.0, 38.0, 26.0, 35.0, 35.0,  nan, 54.0,  2.0, 27.0, 14.0,
              ...
              33.0, 22.0, 28.0, 25.0, 39.0, 27.0, 19.0,  nan, 26.0, 32.0],
             dtype='float64', name='age', length=891)

In [ ]:
#index olarak ekledik, değişken olarak ihtiyacımızın  olmadığını düşünürsek
#age değişkenini sileriz
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"] #indexleri age değişkenine eşitle
df.drop("age", axis=1).head() #age değişkenini sil (axis=1 sütun demektir)

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
age                                                                            
22.0         0       3    male      1      0   7.2500        S  Third    man   
38.0         1       1  female      1      0  71.2833        C  First  woman   
26.0         1       3  female      0      0   7.9250        S  Third  woman   
35.0         1       1  female      1      0  53.1000        S  First  woman   
35.0         0       3    male      0      0   8.0500        S  Third    man   

      adult_male deck  embark_town alive  alone  
age                                              
22.0        True  NaN  Southampton    no  False  
38.0       False    C    Cherbourg   yes  False  
26.0       False  NaN  Southampton   yes   True  
35.0       False    C  Southampton   yes  False  
35.0        True  NaN  Southampton    no   True

In [ ]:
#index olarak ekledik, değişken olarak ihtiyacımızın  olmadığını düşünürsek
#age değişkenini sileriz
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"]
df.drop("age", axis=1, inplace=True) #age değişkenini sil (axis=1 sütun demektir) ,KALICI OLARAK, inplace=True
df.head() #age değişkenini sil (axis=1 sütun demektir)

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
age                                                                            
22.0         0       3    male      1      0   7.2500        S  Third    man   
38.0         1       1  female      1      0  71.2833        C  First  woman   
26.0         1       3  female      0      0   7.9250        S  Third  woman   
35.0         1       1  female      1      0  53.1000        S  First  woman   
35.0         0       3    male      0      0   8.0500        S  Third    man   

      adult_male deck  embark_town alive  alone  
age                                              
22.0        True  NaN  Southampton    no  False  
38.0       False    C    Cherbourg   yes  False  
26.0       False  NaN  Southampton   yes   True  
35.0       False    C  Southampton   yes  False  
35.0        True  NaN  Southampton    no   True

```python
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"]
df.drop("age", axis=1, inplace=True)
df["age"]
```


##### KeyError: 'age'

##### The above exception was the direct cause of the following exception:

##### KeyError    


# Bir index'i değişkene çevirmek

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"]
df.drop("age", axis=1, inplace=True)
df["age"] = df.index #indexleri age değişkenine eşitle 
# df.index'deki değerleri, df["age"] değişkenine atadık , önceden df["age"] yoktur, biz ekledik
# (age değişkeni index olur) 

df.head()  

# son değişken olarak, age , eklendi

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
age                                                                            
22.0         0       3    male      1      0   7.2500        S  Third    man   
38.0         1       1  female      1      0  71.2833        C  First  woman   
26.0         1       3  female      0      0   7.9250        S  Third  woman   
35.0         1       1  female      1      0  53.1000        S  First  woman   
35.0         0       3    male      0      0   8.0500        S  Third    man   

      adult_male deck  embark_town alive  alone   age  
age                                                    
22.0        True  NaN  Southampton    no  False  22.0  
38.0       False    C    Cherbourg   yes  False  38.0  
26.0       False  NaN  Southampton   yes   True  26.0  
35.0       False    C  Southampton   yes  False  35.0  
35.0        True  NaN  Southampton    no   True  35.0

# ikinci yol
# # Bir index'i değişkene çevirmek

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"]
df.drop("age", axis=1, inplace=True)

df.reset_index() #indexleri sıfırla (eski indexler silinir, yeni indexler atanır)

age  survived  pclass     sex  sibsp  parch     fare embarked   class  \
0    22.0         0       3    male      1      0   7.2500        S   Third   
1    38.0         1       1  female      1      0  71.2833        C   First   
2    26.0         1       3  female      0      0   7.9250        S   Third   
3    35.0         1       1  female      1      0  53.1000        S   First   
4    35.0         0       3    male      0      0   8.0500        S   Third   
..    ...       ...     ...     ...    ...    ...      ...      ...     ...   
886  27.0         0       2    male      0      0  13.0000        S  Second   
887  19.0         1       1  female      0      0  30.0000        S   First   
888   NaN         0       3  female      1      2  23.4500        S   Third   
889  26.0         1       1    male      0      0  30.0000        C   First   
890  32.0         0       3    male      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

```python
df.reset_index() 
```
#### 1- index'de yer alan değeri siler
#### 2- bunu sütun olarak(değişken) olarak ekler, kolon olarak


In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.index = df["age"]
df.drop("age", axis=1, inplace=True)

df = df.reset_index() #indexleri sıfırla (eski indexler silinir, yeni indexler atanır)
df

age  survived  pclass     sex  sibsp  parch     fare embarked   class  \
0    22.0         0       3    male      1      0   7.2500        S   Third   
1    38.0         1       1  female      1      0  71.2833        C   First   
2    26.0         1       3  female      0      0   7.9250        S   Third   
3    35.0         1       1  female      1      0  53.1000        S   First   
4    35.0         0       3    male      0      0   8.0500        S   Third   
..    ...       ...     ...     ...    ...    ...      ...      ...     ...   
886  27.0         0       2    male      0      0  13.0000        S  Second   
887  19.0         1       1  female      0      0  30.0000        S   First   
888   NaN         0       3  female      1      2  23.4500        S   Third   
889  26.0         1       1    male      0      0  30.0000        C   First   
890  32.0         0       3    male      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

# Genel olarak, Kolon'da olan bir değeri, index'e gönderdik,
##


# index'te olan değeri ise, 
# değişkene gönderdik

# Satırlarda çalıştık, 
# şimdi kolonlarda, sütünlarda yani
# DEĞİŞKENLERLE ÇALIŞACAĞIZ

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.head() #ilk 5 satırı verir

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.head() #ilk 5 satırı verir


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

# DEĞİŞKENİN VARLIĞINI SORGULAMA
### "age" in df

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
"age" in df

True

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df["age"].head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.age.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

# TİP SORGULAMA

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
type(df.age.head())

pandas.core.series.Series

## Bir değişken seçerken, sonucu ::
# Seri , ya da []
# Dataframe olarak [[]]
# alabilirsiniz

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df[["age"]].head() #age değişkenini verir (dataframe olarak)

age
0  22.0
1  38.0
2  26.0
3  35.0
4  35.0

# Seri:[]

Bir Seri, bir boyutlu bir etiketli dizidir. Bir seri, tek boyutlu bir veri seti saklar ve her bir veri öğesi bir etikete (veya indeks olarak adlandırılır) sahiptir. Bir seri, farklı veri türlerini (tamsayılar, dizeler, kayan noktalı sayılar, Python nesneleri vb.) saklayabilir.

# DataFrame: [[]]

DataFrame, iki boyutlu bir etiketli veri yapısıdır. Bir DataFrame, her biri bir seriyi temsil eden farklı sütunlarla çeşitli veri türlerini (tamsayı, dize, kayan noktalı sayı, Python nesneleri vb.) saklar. Basit bir anlamda, DataFrame birçok Serinin birleşimidir ve bir Excel elektronik tablosuna veya SQL tablosuna benzer şekilde düşünülebilir.

# Farklar

Boyut: En temel fark, Serilerin tek boyutlu ve DataFrame'lerin iki boyutlu veri yapısı olmasıdır.
Veri Saklama: Bir seri, farklı türlerdeki verileri tek bir sütunda saklar. Diğer yandan, bir DataFrame, her biri farklı bir seriyi temsil eden birden çok sütunda veri saklar.
İşlevsellik: DataFrame'ler, verileri düzenlemek, seçmek ve manipüle etmek için daha geniş bir işlevlilik setine sahiptir. DataFrame'ler ayrıca, Serilerde olduğu gibi veriye etiket eklemeye de olanak sağlar ancak aynı zamanda birden fazla indeks (MultiIndex) oluşturmanıza olanak sağlar.


In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
type(df[["age"]].head())

pandas.core.frame.DataFrame

# Birden fazla, DataFrame seçmek
```python
df[["age", "alive"]]

```


In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df[["age", "alive"]]


age alive
0    22.0    no
1    38.0   yes
2    26.0   yes
3    35.0   yes
4    35.0    no
..    ...   ...
886  27.0    no
887  19.0   yes
888   NaN    no
889  26.0   yes
890  32.0    no

[891 rows x 2 columns]

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
col_names = ["age","adult_male","alive"]
df[col_names]

age  adult_male alive
0    22.0        True    no
1    38.0       False   yes
2    26.0       False   yes
3    35.0       False   yes
4    35.0        True    no
..    ...         ...   ...
886  27.0        True    no
887  19.0       False   yes
888   NaN       False    no
889  26.0        True   yes
890  32.0        True    no

[891 rows x 3 columns]

# DataFrame'e yeni bir değişken ekleme

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df["age2"] = df["age"] ** 2 #age değişkeninin karesini al ve age2 değişkenine ata

karsilastir = ["age", "age2"]

df[karsilastir].head() #age ve age2 değişkenlerini verir (dataframe olarak)

age    age2
0  22.0   484.0
1  38.0  1444.0
2  26.0   676.0
3  35.0  1225.0
4  35.0  1225.0


#### .head() kullandığımızda, dataframe'i seriye mi çevirmiş oluyoruz 

#### Hayır, .head() kullanmak DataFrame'i bir Seri'ye çevirmez. .head() fonksiyonu, bir pandas DataFrame'inin veya Seri'nin ilk N satırını döndürür. Varsayılan olarak, .head() ilk 5 satırı döndürür, ancak siz de kaç satır görmek istediğinizi belirleyebilirsiniz.

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df["age2"] = df["age"] ** 2 #age değişkeninin karesini al ve age2 değişkenine ata

karsilastir = ["age", "age2"]

type(df[karsilastir])

pandas.core.frame.DataFrame

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
type(df["age"])

pandas.core.series.Series

#  tek bir köşeli parantez içinde birden fazla sütun adı bulunduğunda, Pandas bir DataFrame döndürür.

#  tek bir köşeli parantez içinde birden fazla sütun adı bulunduğunda, Pandas bir DataFrame döndürür.

#  tek bir köşeli parantez içinde birden fazla sütun adı bulunduğunda, Pandas bir DataFrame döndürür.

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df["age2"] = df["age"] **2
df["age3"] = df["age"] /df["age2"]

df.head()

                         

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone    age2      age3  
0    man        True  NaN  Southampton    no  False   484.0  0.045455  
1  woman       False    C    Cherbourg   yes  False  1444.0  0.026316  
2  woman       False  NaN  Southampton   yes   True   676.0  0.038462  
3  woman       False    C  Southampton   yes  False  1225.0  0.028571  
4    man        True  NaN  Southampton    no   True  1225.0  0.028571

# BİR DEĞİŞKENİ SİLMEK İSTERSEK NE YAPACAĞIZ ? 

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df["age2"] = df["age"] **2
df["age3"] = df["age"] /df["age2"]

df.drop("age3", axis=1) #age3 değişkenini sil (axis=1 sütun demektir)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone    age2  
0      man        True  NaN  Southampton    no  False   484.0  
1    woman       False    C    Cherbourg   yes  False  1444.0  
2    woman       False  NaN  Southampton   yes   True   676.0  
3    woman       False    C  Southampton   yes  False  1225.0  
4      man        True  NaN  Southampton    no   True  1225.0  
..     ...         ...  ...          ...   ...    ...     ...  
886    man        True  NaN  Southampton    no   True   729.0  
887  woman       False    B  Southampton   yes   True   361.0  
888  woman       False  NaN  Southampton    no  False     NaN  
889    man        True    C    Cherbourg   yes   True   676.0  
890    man        True  NaN   Queenstown    no   True  1024.0  

[891 rows x 16 columns]

# BİRDEN FAZLA DEĞİŞKENİ SİLMEK İSTERSEK NE YAPACAĞIZ ? 

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df["age2"] = df["age"] **2
df["age3"] = df["age"] /df["age2"]

col_names =  ["age", "age2", "age3","age","adult_male","alive"]



df.drop(col_names, axis=1) #col_names içindeki değişkenleri sil (axis=1 sütun demektir)

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

    deck  embark_town  alone  
0    NaN  Southampton  False  
1      C    Cherbourg  False  
2    NaN  Southampton   True  
3      C  Southampton  False  
4    NaN  Southampton   True  
..   ...          ...    ...  
886  NaN  Southampton   True  
887    B  Southampton   True  
888  NaN  Southampton  False  
889    C    Cherbourg   True  
890  NaN   Queenstown   True  

[891 rows x 12 columns]

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df["age2"] = df["age"] **2
df["age3"] = df["age"] /df["age2"]

col_names =  ["age", "age2", "age3","age","adult_male","alive"]



df.drop(col_names, axis=1).head()

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
0         0       3    male      1      0   7.2500        S  Third    man   
1         1       1  female      1      0  71.2833        C  First  woman   
2         1       3  female      0      0   7.9250        S  Third  woman   
3         1       1  female      1      0  53.1000        S  First  woman   
4         0       3    male      0      0   8.0500        S  Third    man   

  deck  embark_town  alone  
0  NaN  Southampton  False  
1    C    Cherbourg  False  
2  NaN  Southampton   True  
3    C  Southampton  False  
4  NaN  Southampton   True

In [ ]:
df.loc[0:10, "age"]
# #0. satırdan 10. satıra kadar olan satırları ve age değişkenini verir

0     22.0
1     38.0
2     26.0
3     35.0
4     35.0
5      NaN
6     54.0
7      2.0
8     27.0
9     14.0
10     4.0
Name: age, dtype: float64

In [ ]:
df.loc[:, "age"] #tüm satırları ve age değişkenini verir

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64

In [ ]:
df.loc[:, df.columns.str.contains("age")] 
#tüm satırları ve age içeren değişkenleri verir

age    age2      age3
0    22.0   484.0  0.045455
1    38.0  1444.0  0.026316
2    26.0   676.0  0.038462
3    35.0  1225.0  0.028571
4    35.0  1225.0  0.028571
..    ...     ...       ...
886  27.0   729.0  0.037037
887  19.0   361.0  0.052632
888   NaN     NaN       NaN
889  26.0   676.0  0.038462
890  32.0  1024.0  0.031250

[891 rows x 3 columns]

In [ ]:
df.loc[:, ~df.columns.str.contains("age")] 
#tüm satırları ve age içermeyen değişkenleri verir

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male deck  embark_town alive  alone  
0          True  NaN  Southampton    no  False  
1         False    C    Cherbourg   yes  False  
2         False  NaN  Southampton   yes   True  
3         False    C  Southampton   yes  False  
4          True  NaN  Southampton    no   True  
..          ...  ...          ...   ...    ...  
886        True  NaN  Southampton    no   True  
887       False    B  Southampton   yes   True  
888       False  NaN  Southampton    no  False  
889        True    C    Cherbourg   yes   True  
890        True  NaN   Queenstown    no   True  

[891 rows x 14 columns]

In [ ]:
# iloc ve loc arasındaki fark:
# iloc: index numarası ile seçim yapar
# loc: değişken ismi ile seçim yapar


# iloc ve loc arasındaki fark:
# iloc: index numarası ile seçim yapar
# loc: değişken ismi ile seçim yapar

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.iloc[0:3] 
#0. satırdan 3. satıra kadar olan satırları verir (3. satır dahil değil)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.iloc[0:3, 0:3] 
# 0. satırdan 3. satıra kadar olan satırları ve 0. sütundan 3. sütuna kadar olan sütunları verir (3. satır ve 3. sütun dahil değil)

survived  pclass     sex
0         0       3    male
1         1       1  female
2         1       3  female

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.iloc[0:3, 0:3] 

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.iloc[0, 2] 
#0. satır ve 2. sütunun değerini verir

'male'

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.loc[0:3]  
#0. satırdan 3. satıra kadar olan satırları verir (3. satır dahil değil)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False

```python
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.iloc[0:3, "age"]

ValueError: Can only index by location with a [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array]

The above exception was the direct cause of the following exception:
```

# ----------------------------------------------------
# ----------------------------------------------------
# ----------------------------------------------------
# ----------------------------------------------------
# ----------------------------------------------------
# ----------------------------------------------------
# ----------------------------------------------------
# ----------------------------------------------------


# iloc , integer based , bir sorgu aracı
# string ifade ile arayamazsın

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df.loc[0:3, "age"]

0    22.0
1    38.0
2    26.0
3    35.0
Name: age, dtype: float64

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

col_names = ["age","embarked","alive"]
df.loc[0:3, col_names]

age embarked alive
0  22.0        S    no
1  38.0        C   yes
2  26.0        S   yes
3  35.0        S   yes

```python
df.loc[0:3, col_names]
``` 

#### kullanıldığında, burada loc kullanılıyor ve loc etiket tabanlı bir işlem olduğu için, belirttiğiniz 3 etiketli satır dahil olur. Yani bu kod 0, 1, 2 ve 3 etiketli satırları seçer.

```python
df.iloc[0:3, col_names]
``` 

#### bu durumda indeks tabanlı olduğu için 3 indeksi dahil etmezdi ve sadece 0, 1 ve 2 indeksli satırları seçerdi.


belirttiğim gibi loc ile yaptığınız dilimlemelerde bitiş indeksi dahildir.

# KOŞULLU SEÇİM İŞLEMLERİ

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df[df["age"] > 50].head() #age değişkeni 50'dan büyük olan satırları verir

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
6          0       1    male  54.0      0      0  51.8625        S   First   
11         1       1  female  58.0      0      0  26.5500        S   First   
15         1       2  female  55.0      0      0  16.0000        S  Second   
33         0       2    male  66.0      0      0  10.5000        S  Second   
54         0       1    male  65.0      0      1  61.9792        C   First   

      who  adult_male deck  embark_town alive  alone  
6     man        True    E  Southampton    no   True  
11  woman       False    C  Southampton   yes   True  
15  woman       False  NaN  Southampton   yes   True  
33    man        True  NaN  Southampton    no   True  
54    man        True    B    Cherbourg    no  False

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df[df["age"] > 50].count() #age değişkeni 50'dan büyük olan satır sayısını verir

survived       64
pclass         64
sex            64
age            64
sibsp          64
parch          64
fare           64
embarked       63
class          64
who            64
adult_male     64
deck           33
embark_town    63
alive          64
alone          64
dtype: int64

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")
df[df["age"] > 50]["age"].count() 
#age değişkeni 50'dan büyük olan satır sayısını verir 
#(sadece age değişkenini verir)


64

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df.loc[df["age"] > 50 , "class"].head()
#age değişkeni 50'dan büyük olan satırları ve class değişkenini verir

6      First
11     First
15    Second
33    Second
54     First
Name: class, dtype: category
Categories (3, object): ['First', 'Second', 'Third']

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df.loc[df["age"] > 50 , ["age","class"]].head()

age   class
6   54.0   First
11  58.0   First
15  55.0  Second
33  66.0  Second
54  65.0   First

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df.loc[(df["age"] > 50) & (df["sex"] =="male") , ["age","class"]].head() 


age   class
6   54.0   First
33  66.0  Second
54  65.0   First
94  59.0   Third
96  71.0   First

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df.loc[(df["age"] > 50) & (df["sex"] =="male") & (df["embark_town"] == "Cherbourg") , ["age","class"]].head() # 3 koşul var (and ile bağlıyoruz)

age  class
54   65.0  First
96   71.0  First
155  51.0  First
174  56.0  First
487  58.0  First

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df.loc[(df["age"] > 50) 
       & (df["sex"] =="male") 
       & (df["embark_town"] == "Cherbourg") , ["age","class","embark_town"]].head() # 3 koşul var (and ile bağlıyoruz)

age  class embark_town
54   65.0  First   Cherbourg
96   71.0  First   Cherbourg
155  51.0  First   Cherbourg
174  56.0  First   Cherbourg
487  58.0  First   Cherbourg

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df.loc[(df["age"] > 50) 
       & (df["sex"] =="male") 
       & ((df["embark_town"] == "Cherbourg") | (df["embark_town"] == "Southampton")) , 
       ["age","class","embark_town"]].head() # 3 koşul var (and ile bağlıyoruz)

age   class  embark_town
6   54.0   First  Southampton
33  66.0  Second  Southampton
54  65.0   First    Cherbourg
94  59.0   Third  Southampton
96  71.0   First    Cherbourg

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df["embark_town"].value_counts()

Southampton    644
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64

## value_counts() fonksiyonu, pandas kütüphanesinde bir serinin (veya bir DataFrame'in tek bir sütununun) benzersiz değerlerinin sayısını hesaplamak için kullanılır.

In [ ]:
import pandas as pd
import seaborn as sns

pd.set_option('display.max_columns', None) #tüm sütunları göster

df = sns.load_dataset("titanic")

df_new = df.loc[(df["age"] > 50) 
       & (df["sex"] =="male") 
       & ((df["embark_town"] == "Cherbourg") | (df["embark_town"] == "Southampton")) , 
       ["age","class","embark_town"]].head() # 3 koşul var (and ile bağlıyoruz)

df_new["embark_town"].value_counts()



Southampton    3
Cherbourg      2
Name: embark_town, dtype: int64

# Topluluaştırma ve Gruplama (Aggregation & Grouping)

# count() 
# first()
# last()
# mean()
# median()
# min()
# max()
# std()

In [ ]:
# count() : veri setindeki gözlem sayısını verir (satır sayısı)  (boş değerler hariç)
# first() : veri setindeki ilk gözlemi verir (ilk satırı verir) 
# last() : veri setindeki son gözlemi verir (son satırı verir) 
# mean() : veri setindeki sayısal değişkenlerin ortalamasını verir
# median() : veri setindeki sayısal değişkenlerin medyanını verir
# min() : veri setindeki sayısal değişkenlerin minimum değerini verir
# max() : veri setindeki sayısal değişkenlerin maksimum değerini verir
# std() : veri setindeki sayısal değişkenlerin standart sapmasını verir
# var() : veri setindeki sayısal değişkenlerin varyansını verir
# sum() : veri setindeki sayısal değişkenlerin toplamını verir

# pivot table : veri setindeki kategorik değişkenlerin ortalamasını verir



In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")


df.head() #ilk 5 satırı verir

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

# kadınların ve erkeklerin yaş ortalamasına erişmek istiyorum.
df["age"]

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64

In [ ]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

# kadınların ve erkeklerin yaş ortalamasına erişmek istiyorum.
df["age"].mean() #tüm veri setindeki yaş ortalamasını verir

29.69911764705882

In [157]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

# kadınların ve erkeklerin yaş ortalamasına erişmek istiyorum.
df.groupby("sex")

In [159]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

# kadınların ve erkeklerin yaş ortalamasına erişmek istiyorum.
df.groupby("sex")["age"].mean() #cinsiyete göre yaş ortalamasını verir

sex
female    27.915709
male      30.726645
Name: age, dtype: float64

In [160]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

# kadınların ve erkeklerin yaş ortalamasına erişmek istiyorum.
df.groupby("sex").agg({"age":"mean"}) #cinsiyete göre yaş ortalamasını verir
# agg : aggregate (toplamak, bir araya getirmek)


age
sex              
female  27.915709
male    30.726645

In [162]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

# kadınların ve erkeklerin yaş ortalamasına erişmek istiyorum.
df.groupby("sex").agg({"age":["mean","sum"]}) 
#cinsiyete göre yaş ortalamasını ve toplamını verir

age          
             mean       sum
sex                        
female  27.915709   7286.00
male    30.726645  13919.17

In [166]:
df.groupby("sex").agg({"age":["mean","sum"],
                       "embark_town":"count"}) 
#cinsiyete göre yaş ortalamasını ve toplamını verir ve embark_town değişkeninin sayısını verir


age           embark_town
             mean       sum       count
sex                                    
female  27.915709   7286.00         312
male    30.726645  13919.17         577

### emberk'ın frekansını vermedi, cinsiyetin frekansını verdi


In [167]:
df.groupby("sex").agg({"age":["mean","sum"],
                       "survived":"mean"}) 
#cinsiyete göre yaş ortalamasını ve toplamını verir ve survived değişkeninin ortalamasını verir

age            survived
             mean       sum      mean
sex                                  
female  27.915709   7286.00  0.742038
male    30.726645  13919.17  0.188908

Bu kod parçacığı, veri setini 'sex' sütununa göre gruplandırır. Yani, tüm analizler ve hesaplamalar cinsiyete göre ayrılmış gruplar üzerinde gerçekleştirilir.

Ardından 'agg' fonksiyonu ile 'age' sütunu için ortalama ('mean') ve toplam ('sum') hesaplanır, 'survived' sütunu içinse ortalama ('mean') hesaplanır.

Özetle, bu kod parçacığı cinsiyetlere göre yaş ortalamalarını, yaş toplamlarını ve hayatta kalma oranlarını hesaplar.

Hayatta kalma oranı genellikle 0 ve 1 değerlerini alır (0 - hayatta kalmadı, 1 - hayatta kaldı). Bu durumda, 'survived' sütununun ortalama değeri, o cinsiyet grubunun hayatta kalma oranını temsil eder. Örneğin, kadınların hayatta kalma oranı, 'survived' sütununun kadınlar için ortalama değeri olarak hesaplanır.

# Yaşa göre kırılım yaptık, 
# şimdi cinsiye göre kırılım ve , sonra Limana binişe göre kırılım yapalım

In [169]:
df.groupby(["sex","embark_town"]).agg({"age":["mean","sum"],
                       "survived":"mean"}) 

age            survived
                         mean       sum      mean
sex    embark_town                               
female Cherbourg    28.344262   1729.00  0.876712
       Queenstown   24.291667    291.50  0.750000
       Southampton  27.771505   5165.50  0.689655
male   Cherbourg    32.998841   2276.92  0.305263
       Queenstown   30.937500    495.00  0.073171
       Southampton  30.291440  11147.25  0.174603

In [170]:
df.groupby(["sex","embark_town","class"]).agg({"age":["mean","sum"],
                       "survived":"mean"}) 

age           survived
                                mean      sum      mean
sex    embark_town class                               
female Cherbourg   First   36.052632  1370.00  0.976744
                   Second  19.142857   134.00  1.000000
                   Third   14.062500   225.00  0.652174
       Queenstown  First   33.000000    33.00  1.000000
                   Second  30.000000    30.00  1.000000
                   Third   22.850000   228.50  0.727273
       Southampton First   32.704545  1439.00  0.958333
                   Second  29.719697  1961.50  0.910448
                   Third   23.223684  1765.00  0.375000
male   Cherbourg   First   40.111111  1444.00  0.404762
                   Second  25.937500   207.50  0.200000
                   Third   25.016800   625.42  0.232558
       Queenstown  First   44.000000    44.00  0.000000
                   Second  57.000000    57.00  0.000000
                   Third   28.142857   394.00  0.076923
       Southampton First   41.897188  2681.42  0.354430
                   Second  30.875889  2778.83  0.154639
                   Third   26.574766  5687.00  0.128302

In [171]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

df.groupby(["sex","embark_town","class"]).agg({
    "age":["mean","sum"],
    "survived":"mean",
    "sex":"count"}) 

age           survived   sex
                                mean      sum      mean count
sex    embark_town class                                     
female Cherbourg   First   36.052632  1370.00  0.976744    43
                   Second  19.142857   134.00  1.000000     7
                   Third   14.062500   225.00  0.652174    23
       Queenstown  First   33.000000    33.00  1.000000     1
                   Second  30.000000    30.00  1.000000     2
                   Third   22.850000   228.50  0.727273    33
       Southampton First   32.704545  1439.00  0.958333    48
                   Second  29.719697  1961.50  0.910448    67
                   Third   23.223684  1765.00  0.375000    88
male   Cherbourg   First   40.111111  1444.00  0.404762    42
                   Second  25.937500   207.50  0.200000    10
                   Third   25.016800   625.42  0.232558    43
       Queenstown  First   44.000000    44.00  0.000000     1
                   Second  57.000000    57.00  0.000000     1
                   Third   28.142857   394.00  0.076923    39
       Southampton First   41.897188  2681.42  0.354430    79
                   Second  30.875889  2778.83  0.154639    97
                   Third   26.574766  5687.00  0.128302   265

# PİVOT TABLE 

Veri setini kırılımlar açısından değerlendirir ve
İlgilendiğimiz özet istatistiği , 
bu kırılımlar açısından görmemizi sağlar...


Yaş ve Gemiye binme, Lokasyonunu ifade eden, 
bu iki değişken açısından, bir pivot tablo oluşturmak istediğimizi düşünelim...
ve bunların kesişimine de yaş bilgisine erişmek istediğimizi düşünelim ya da survived bilgisine erişmek istediğimi düşünelim

```python
df.pivot_table("1 kesişimde neyi görmek istiyorsan onu gir,
                2 index'de satırda hangi değişkeni görmek istiyorsun,
                3 sütunda hangi değişkeni görmek istiyorsun
                      )
```

In [173]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.pivot_table("survived","sex","embarked") #survived değişkeninin ortalamasını verir
# default ta aggfunc = "mean" dir (ortalama alır) 

embarked         C         Q         S
sex                                   
female    0.876712  0.750000  0.689655
male      0.305263  0.073171  0.174603

In [174]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.pivot_table("survived","sex","embarked",aggfunc="std")  #survived değişkeninin standart sapmasını verir

embarked         C         Q         S
sex                                   
female    0.331042  0.439155  0.463778
male      0.462962  0.263652  0.380058

In [176]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df.pivot_table("survived","sex",["embarked","class"]) 
#survived değişkeninin ortalamasını verir
# embarked ve class değişkenlerini gruplar ve ortalamasını alır


embarked         C                      Q                          S  \
class        First Second     Third First Second     Third     First   
sex                                                                    
female    0.976744    1.0  0.652174   1.0    1.0  0.727273  0.958333   
male      0.404762    0.2  0.232558   0.0    0.0  0.076923  0.354430   

embarked                      
class       Second     Third  
sex                           
female    0.910448  0.375000  
male      0.154639  0.128302

Yaşlara göre de analiz yapmak istiyorum ama Yaş, integer değer olduğu için
```python
df["new_age"] = pd.cut(df["age"],[0,10,18,25,40,90]) #hangi kategöriye böleceğimi biliyorsam
df["new_age"] = pd.qcut() # bilmiyorsam
```

In [178]:
df["new_age"] = pd.cut(df["age"],[0,10,18,25,40,90]) 
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone   new_age  
0    man        True  NaN  Southampton    no  False  (18, 25]  
1  woman       False    C    Cherbourg   yes  False  (25, 40]  
2  woman       False  NaN  Southampton   yes   True  (25, 40]  
3  woman       False    C  Southampton   yes  False  (25, 40]  
4    man        True  NaN  Southampton    no   True  (25, 40]

Yaş kırılımında da , hayatta kalanları inceleyelim

In [179]:
df.pivot_table("survived","sex","new_age")
# ("kesişimHangiDeğer, "Satır Ne olmasını ?", "Sütün ne olmasını?")
# survived değişkeninin ortalamasını verir

new_age   (0, 10]  (10, 18]  (18, 25]  (25, 40]  (40, 90]
sex                                                      
female   0.612903  0.729730  0.759259  0.802198  0.770833
male     0.575758  0.131579  0.120370  0.220930  0.176471

Bir boyut daha ekleyelim , Yolculuk sınıfını ve bunun kategorik sınıfını ekleyelim

In [180]:
df.pivot_table("survived","sex",["new_age","class"]) 
# ("kesişimHangiDeğer, "Satır Ne olmasını ?", "Sütün ne olmasını?")


new_age (0, 10]                   (10, 18]                   (18, 25]  \
class     First Second     Third     First Second     Third     First   
sex                                                                     
female      0.0    1.0  0.500000  1.000000    1.0  0.523810  0.941176   
male        1.0    1.0  0.363636  0.666667    0.0  0.103448  0.333333   

new_age                      (25, 40]                      (40, 90]            \
class      Second     Third     First    Second     Third     First    Second   
sex                                                                             
female   0.933333  0.500000  1.000000  0.906250  0.464286  0.961538  0.846154   
male     0.047619  0.115385  0.513514  0.071429  0.172043  0.280000  0.095238   

new_age            
class       Third  
sex                
female   0.111111  
male     0.064516

In [182]:
pd.set_option("display.width",500) #ekran genişliğini ayarla (500 piksel)
df.pivot_table("survived","sex",["new_age","class"]) 

new_age (0, 10]                   (10, 18]                   (18, 25]                      (25, 40]                      (40, 90]                    
class     First Second     Third     First Second     Third     First    Second     Third     First    Second     Third     First    Second     Third
sex                                                                                                                                                  
female      0.0    1.0  0.500000  1.000000    1.0  0.523810  0.941176  0.933333  0.500000  1.000000  0.906250  0.464286  0.961538  0.846154  0.111111
male        1.0    1.0  0.363636  0.666667    0.0  0.103448  0.333333  0.047619  0.115385  0.513514  0.071429  0.172043  0.280000  0.095238  0.064516

# apply
satır ya da sütunlarda , otomatik olarak fonksiyon çalıştırma imkanı sağlar

### bir data frame'e ""apply"" ile istediğimiz fonksiyonu uygulayabiliriz
### satır ya da sütunlarda

# lambda
### bir fonksiyon tanımlama şeklidir, def gibi
### farkı, kullan at fonksiyondur

In [184]:
df["age2"] = df["age"] *2
df["age3"] = df["age"] *5
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town alive  alone   new_age  age2   age3
0         0       3    male  22.0      1      0   7.2500        S  Third    man        True  NaN  Southampton    no  False  (18, 25]  44.0  110.0
1         1       1  female  38.0      1      0  71.2833        C  First  woman       False    C    Cherbourg   yes  False  (25, 40]  76.0  190.0
2         1       3  female  26.0      0      0   7.9250        S  Third  woman       False  NaN  Southampton   yes   True  (25, 40]  52.0  130.0
3         1       1  female  35.0      1      0  53.1000        S  First  woman       False    C  Southampton   yes  False  (25, 40]  70.0  175.0
4         0       3    male  35.0      0      0   8.0500        S  Third    man        True  NaN  Southampton    no   True  (25, 40]  70.0  175.0

Bu veri seti içerisindeki yaş, değişkenlerinin 10'a bölünmesini istiyoruz

In [185]:
df["age"]/10
df["age2"]/10 
df["age3"]/10

0      11.0
1      19.0
2      13.0
3      17.5
4      17.5
       ... 
886    13.5
887     9.5
888     NaN
889    13.0
890    16.0
Name: age3, Length: 891, dtype: float64

```python
df["age"]/10
```


  Cell In[186], line 1
    ```python
    ^
SyntaxError: invalid syntax

In [192]:
(df["age"]/10).head()

0    2.2
1    3.8
2    2.6
3    3.5
4    3.5
Name: age, dtype: float64

In [193]:
(df["age2"]/10).head()

0    4.4
1    7.6
2    5.2
3    7.0
4    7.0
Name: age2, dtype: float64

In [194]:
(df["age3"]/10).head()

0    11.0
1    19.0
2    13.0
3    17.5
4    17.5
Name: age3, dtype: float64

In [195]:
for col in df.columns:
    print(col)

survived
pclass
sex
age
sibsp
parch
fare
embarked
class
who
adult_male
deck
embark_town
alive
alone
new_age
age2
age3


In [196]:
for col in df.columns:
    if "age" in col:
        print(col)


age
new_age
age2
age3


In [198]:
for col in df.columns:
    if "age" in col and col != "new_age":
        print(df[col]/10) #age içeren değişkenlerin değerlerini 10'a böl 

0      2.2
1      3.8
2      2.6
3      3.5
4      3.5
      ... 
886    2.7
887    1.9
888    NaN
889    2.6
890    3.2
Name: age, Length: 891, dtype: float64
0      4.4
1      7.6
2      5.2
3      7.0
4      7.0
      ... 
886    5.4
887    3.8
888    NaN
889    5.2
890    6.4
Name: age2, Length: 891, dtype: float64
0      11.0
1      19.0
2      13.0
3      17.5
4      17.5
       ... 
886    13.5
887     9.5
888     NaN
889    13.0
890    16.0
Name: age3, Length: 891, dtype: float64


In [199]:
for col in df.columns:
    if "age" in col and col != "new_age":
        print((df[col]/10).head()) #age içeren değişkenlerin değerlerini 10'a böl 

0    2.2
1    3.8
2    2.6
3    3.5
4    3.5
Name: age, dtype: float64
0    4.4
1    7.6
2    5.2
3    7.0
4    7.0
Name: age2, dtype: float64
0    11.0
1    19.0
2    13.0
3    17.5
4    17.5
Name: age3, dtype: float64


In [200]:
for col in df.columns:
    if "age" in col and col != "new_age":
        df[col] = df[col] / 10


In [201]:
for col in df.columns:
    if "age" in col and col != "new_age":
        df[col] = df[col] / 10

df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town alive  alone   new_age  age2  age3
0         0       3    male  0.22      1      0   7.2500        S  Third    man        True  NaN  Southampton    no  False  (18, 25]  0.44  1.10
1         1       1  female  0.38      1      0  71.2833        C  First  woman       False    C    Cherbourg   yes  False  (25, 40]  0.76  1.90
2         1       3  female  0.26      0      0   7.9250        S  Third  woman       False  NaN  Southampton   yes   True  (25, 40]  0.52  1.30
3         1       1  female  0.35      1      0  53.1000        S  First  woman       False    C  Southampton   yes  False  (25, 40]  0.70  1.75
4         0       3    male  0.35      0      0   8.0500        S  Third    man        True  NaN  Southampton    no   True  (25, 40]  0.70  1.75

In [202]:
df[["age","age2","age3"]].apply(lambda x: x/10).head() #age, age2, age3 değişkenlerinin değerlerini 10'a böl

age   age2   age3
0  0.022  0.044  0.110
1  0.038  0.076  0.190
2  0.026  0.052  0.130
3  0.035  0.070  0.175
4  0.035  0.070  0.175

In [204]:
df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(lambda x: x/10).head()
 # :, bütün satırları seç
 # df.columns.str.contains("age") , age içeren değişkenleri seç
    # df.columns != 'new_age' , new_age değişkenini seçme
    # & , and demektir
    # df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')] , age içeren değişkenleri seç ve new_age değişkenini seçme
    # df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(lambda x: x/10) , age içeren değişkenlerin değerlerini 10'a böl
    # df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(lambda x: x/10).head() , age içeren değişkenlerin değerlerini 10'a böl ve ilk 5 satırı ver


age   age2   age3
0  0.022  0.044  0.110
1  0.038  0.076  0.190
2  0.026  0.052  0.130
3  0.035  0.070  0.175
4  0.035  0.070  0.175

In [205]:
df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(lambda x: (x-x.mean())/ x.std()).head()   
# age içeren değişkenlerin değerlerini standartlaştır (z-score) ve ilk 5 satırı ver 


age      age2      age3
0 -0.530005 -0.530005 -0.530005
1  0.571430  0.571430  0.571430
2 -0.254646 -0.254646 -0.254646
3  0.364911  0.364911  0.364911
4  0.364911  0.364911  0.364911

In [206]:
def standart_scaler(col_name): #standartlaştırma fonksiyonu
    return (col_name - col_name.mean()) / col_name.std() #z-score formülü 

df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(standart_scaler).head()   


age      age2      age3
0 -0.530005 -0.530005 -0.530005
1  0.571430  0.571430  0.571430
2 -0.254646 -0.254646 -0.254646
3  0.364911  0.364911  0.364911
4  0.364911  0.364911  0.364911

In [208]:
def standart_scaler(col_name): #standartlaştırma fonksiyonu
    return (col_name - col_name.mean()) / col_name.std() #z-score formülü 

df.loc[:, ["age","age2","age3"] ] = df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(standart_scaler).head()  
df.head() 


survived  pclass     sex       age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town alive  alone   new_age      age2      age3
0         0       3    male -1.346261      1      0   7.2500        S  Third    man        True  NaN  Southampton    no  False  (18, 25] -1.346261 -1.346261
1         1       1  female  0.995063      1      0  71.2833        C  First  woman       False    C    Cherbourg   yes  False  (25, 40]  0.995063  0.995063
2         1       3  female -0.760930      0      0   7.9250        S  Third  woman       False  NaN  Southampton   yes   True  (25, 40] -0.760930 -0.760930
3         1       1  female  0.556064      1      0  53.1000        S  First  woman       False    C  Southampton   yes  False  (25, 40]  0.556064  0.556064
4         0       3    male  0.556064      0      0   8.0500        S  Third    man        True  NaN  Southampton    no   True  (25, 40]  0.556064  0.556064

In [209]:
def standart_scaler(col_name): #standartlaştırma fonksiyonu
    return (col_name - col_name.mean()) / col_name.std() #z-score formülü 

df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age') ] = df.loc[:, df.columns.str.contains("age") & (df.columns != 'new_age')].apply(standart_scaler).head()  
df.head() 

survived  pclass     sex       age  sibsp  parch     fare embarked  class    who  adult_male deck  embark_town alive  alone   new_age      age2      age3
0         0       3    male -1.346261      1      0   7.2500        S  Third    man        True  NaN  Southampton    no  False  (18, 25] -1.346261 -1.346261
1         1       1  female  0.995063      1      0  71.2833        C  First  woman       False    C    Cherbourg   yes  False  (25, 40]  0.995063  0.995063
2         1       3  female -0.760930      0      0   7.9250        S  Third  woman       False  NaN  Southampton   yes   True  (25, 40] -0.760930 -0.760930
3         1       1  female  0.556064      1      0  53.1000        S  First  woman       False    C  Southampton   yes  False  (25, 40]  0.556064  0.556064
4         0       3    male  0.556064      0      0   8.0500        S  Third    man        True  NaN  Southampton    no   True  (25, 40]  0.556064  0.556064